src/haskell/parser/src/KoreAST.hs

```diff
@@ -3,119 +3,163 @@ module KoreAST where
 newtype Id = Id String
   deriving (Show, Eq)
 
+newtype MetaId = MetaId String
+  deriving (Show, Eq)
+
 newtype StringLiteral = StringLiteral { getStringLiteral :: String }
     deriving (Show, Eq)
 
-data SymbolOrAlias = SymbolOrAlias
-    { symbolOrAliasConstructor :: !Id
-    , symbolOrAliasParams      :: ![Sort]
+data ObjectSymbolOrAlias = ObjectSymbolOrAlias
+    { objectSymbolOrAliasConstructor :: !Id
+    , objectSymbolOrAliasParams      :: ![ObjectSort]
+    }
+    deriving (Show, Eq)
+
+data ObjectSymbol = ObjectSymbol
+    { objectSymbolConstructor :: !Id
+    , objectSymbolParams      :: ![ObjectSort]
     }
     deriving (Show, Eq)
 
-data Symbol = Symbol
-    { symbolConstructor :: !Id
-    , symbolParams      :: ![Sort]
+data ObjectAlias = ObjectAlias
+    { objectAliasConstructor :: !Id
+    , objectAliasParams      :: ![ObjectSort]
     }
     deriving (Show, Eq)
 
-data Alias = Alias
-    { aliasConstructor :: !Id
-    , aliasParams      :: ![Sort]
+data MetaSymbolOrAlias = MetaSymbolOrAlias
+    { metaSymbolOrAliasConstructor :: !MetaId
+    , metaSymbolOrAliasParams      :: ![MetaSort]
     }
     deriving (Show, Eq)
 
-newtype SortVariable = SortVariable { getSortVariable :: Id }
+data MetaSymbol = MetaSymbol
+    { metaSymbolConstructor :: !MetaId
+    , metaSymbolParams      :: ![MetaSort]
+    }
+    deriving (Show, Eq)
+
+data MetaAlias = MetaAlias
+    { metaAliasConstructor :: !MetaId
+    , metaAliasParams      :: ![MetaSort]
+    }
     deriving (Show, Eq)
 
-data Sort
-    = SortVariableSort !SortVariable
+newtype ObjectSortVariable = ObjectSortVariable { getObjectSortVariable :: Id }
+    deriving (Show, Eq)
+
+data ObjectSort
+    = ObjectSortVariableSort !ObjectSortVariable
     | ActualSort
         { actualSortName  :: !Id
-        , actualSortSorts :: ![Sort]
+        , actualSortSorts :: ![ObjectSort]
         }
     deriving (Show, Eq)
 
+newtype MetaSortVariable = MetaSortVariable { getMetaSortVariable :: MetaId }
+    deriving (Show, Eq)
+
+data SortVariable
+    = ObjectSortVariableSortVariable ObjectSortVariable
+    | MetaSortVariableSortVariable MetaSortVariable
+    deriving (Show, Eq)
+
+data MetaSort
+    = MetaSortVariableMetaSort !MetaSortVariable
+    | CharMetaSort
+    | CharListMetaSort
+    | PatternMetaSort
+    | PatternListMetaSort
+    | SortMetaSort
+    | SortListMetaSort
+    | StringMetaSort
+    | SymbolMetaSort
+    | SymbolListMetaSort
+    | VariableMetaSort
+    | VariableListMetaSort
+    deriving (Show, Eq)
+
 newtype ModuleName = ModuleName { getModuleName :: String }
     deriving (Show, Eq)
 
 data Variable = Variable
     { variableName :: !Id
-    , variableSort :: !Sort
+    , variableSort :: !ObjectSort
     }
     deriving (Show, Eq)
 
 data Pattern
     = AndPattern
-        { andPatternSort   :: !Sort
+        { andPatternSort   :: !ObjectSort
         , andPatternFirst  :: !Pattern
         , andPatternSecond :: !Pattern
         }
     | ApplicationPattern
```

Comment:
Combining sums with records is somewhat frowned upon (see https://stackoverflow.com/questions/37652243/record-syntax-and-sum-types for details).